pkg/cloudsql/adapter.go

```diff
@@ -0,0 +1,263 @@
+
+	flags := []*sqladmin.DatabaseFlags{}
+
+	for _, knob := range proposedConfig.KnobsOverrides {
+		knobConfig, err := parameters.FindRecommendedKnob(proposedConfig.Config, knob)
+		if err != nil {
+			return fmt.Errorf("failed to find recommended knob: %v", err)
+		}
+		fmtValue, err := knobConfig.GetSettingValue()
+		if err != nil {
+			return fmt.Errorf("failed to get setting value: %v", err)
+		}
+
+		param := &sqladmin.DatabaseFlags{
+			Name:  knobConfig.Name,
+			Value: fmtValue,
+		}
+
+		flags = append(flags, param)
+	}
+
+	adapter.CloudSQLAdminClient.applyFlags(adapter.CloudSQLConfig.ProjectID, adapter.CloudSQLConfig.DatabaseName, flags)
+
+	return nil
+}
+
+func (adapter *CloudSQLAdapter) GetActiveConfig() (agent.ConfigArraySchema, error) {
+	adapter.Logger().Debugf("Getting Active Config")
+
+	return pg.GetActiveConfig(adapter.PGDriver, context.Background(), adapter.Logger())
+}
+
+func (adapter *CloudSQLAdapter) GetSystemInfo() ([]metrics.FlatValue, error) {
+	adapter.Logger().Debugf("Getting System Info")
+
+	// Get PostgreSQL version and max connections from database
+	pgVersion, err := pg.PGVersion(adapter.PGDriver)
+	if err != nil {
+		return nil, err
+	}
+
+	maxConnections, err := pg.MaxConnections(adapter.PGDriver)
+	if err != nil {
+		return nil, err
+	}
+
+	version, err := metrics.PGVersion.AsFlatValue(pgVersion)
+	if err != nil {
+		adapter.Logger().Errorf("Error creating PostgreSQL version metric: %v", err)
+		return nil, err
+	}
+
+	maxConnectionsMetric, err := metrics.PGMaxConnections.AsFlatValue(maxConnections)
+	if err != nil {
+		adapter.Logger().Errorf("Error creating max connections metric: %v", err)
+		return nil, err
+	}
+
+	cpuCount, err := GetCPUCount(adapter.CloudMonitoringClient, adapter.CloudSQLConfig.ProjectID, adapter.CloudSQLConfig.DatabaseName)
+	if err != nil {
+		adapter.Logger().Errorf("Error fetching CPU count: %v", err)
+		return nil, err
+	}
+
+	cpuCountMetric, err := metrics.NodeCPUCount.AsFlatValue(int64(cpuCount.GetDoubleValue()))
+	if err != nil {
+		adapter.Logger().Errorf("Error creating NodeCPUCount metric: %v", err)
+		return nil, err
+	}
+
+	memoryTotal, err := GetMemoryTotal(adapter.CloudMonitoringClient, adapter.CloudSQLConfig.ProjectID, adapter.CloudSQLConfig.DatabaseName)
+	if err != nil {
+		adapter.Logger().Errorf("failed to get memory total: %v", err)
+		return nil, err
+	}
+
+	memoryTotalMetric, err := metrics.NodeMemoryTotal.AsFlatValue(memoryTotal.GetInt64Value())
+	if err != nil {
+		adapter.Logger().Errorf("Error creating NodeMemoryTotal metric: %v", err)
+		return nil, err
+	}
+
+	systemInfo := []metrics.FlatValue{
+		version,
+		maxConnectionsMetric,
+		cpuCountMetric,
+		memoryTotalMetric,
+	}
+
+	adapter.Logger().Debugf("SystemMetrics: %v", systemInfo)
+
+	return systemInfo, nil
+}
+
+func (adapter *CloudSQLAdapter) Guardrails() *guardrails.Signal {
+	if time.Since(adapter.State.LastGuardrailCheck) < 5*time.Second {
+		return nil
+	}
```

Comment:
This is only needed if it doesn't make sense to tick at the time defined in the main agent loop, which is set to 15 seconds. What you might see in other adapaters is other times based on when the last memory was fetched. In the case of Aiven, this was because they didn't give better than 30 second resolution and it made no sense to spam their API. Their API also returned a *lot* of data, for when we only wanted one value from it.

https://github.com/dbtuneai/dbtune-agent/blob/64c133b027916520d606cdb50a84262f09251699/pkg/runner/runner.go#L88-L107